### PR TITLE
feat(instrumentation): new spans for llms

### PR DESCRIPTION
**Pull request by RogerHYang**
## Before
Screenshot 2024-05-17 at 7 17 43 AM

## After
Screenshot 2024-05-17 at 7 39 40 AM

## Code

```python
def multiply(a: int, b: int) -> int:
    """Multiple two integers and returns the result integer"""
    return a * b


def add(a: int, b: int) -> int:
    """Add two integers and returns the result integer"""
    return a + b


multiply_tool = FunctionTool.from_defaults(fn=multiply)
add_tool = FunctionTool.from_defaults(fn=add)
agent = OpenAIAgent.from_tools([multiply_tool, add_tool])
Settings.llm = OpenAI(model="gpt-3.5-turbo")

if __name__ == "__main__":
    response = agent.query("What is (121 * 3) + 42?")
    print(response)
```

### CODE REVIEW

File: llama-index-core/llama_index/core/base/llms/base.py [hunk 1]
```diff
@@ -253,3 +254,19 @@ async def astream_complete(
                 print(response.text, end="", flush=True)
             ```
         """
+
+    def __init_subclass__(cls, **kwargs) -> None:
```
Comment (nerdai) [inline llama-index-core/llama_index/core/base/llms/base.py:258]: Ohhhh this is nice!

Up to now, I've been resorting to decorating the subclasses individually. If I'm reading this correctly, then this new hook allows for "decorating" these methods with span automatically.

Comment (RogerHYang) [reply llama-index-core/llama_index/core/base/llms/base.py:258]: Yes, all subclasses will have these methods decorated automatically.

File: llama-index-core/llama_index/core/llms/callbacks.py [hunk 2]
```diff
@@ -209,6 +209,19 @@ def dummy_wrapper(_self: Any, *args: Any, **kwargs: Any) -> Any:
         if not is_wrapped:
             f.__wrapped__ = True  # type: ignore
 
+        for attr in (
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/llms/callbacks.py:212]: Can we add a comment for what this is doing? I actually don't quite get what is happening here haha

File: llama-index-core/llama_index/core/base/llms/base.py (hunk 1 above)
Comment (logan-markewich) [inline llama-index-core/llama_index/core/base/llms/base.py:258]: Lets add a docstring to this method to explain what's happening in here for future devs

Comment (logan-markewich) [reply llama-index-core/llama_index/core/base/llms/base.py:258]: Lets add a docstring to this method to explain what's happening in here for future devs

Comment (RogerHYang) [reply llama-index-core/llama_index/core/base/llms/base.py:258]: added

File: llama-index-core/llama_index/core/llms/callbacks.py (hunk 2 above)
Comment (RogerHYang) [reply llama-index-core/llama_index/core/llms/callbacks.py:214]: Added. 

Screenshots below show the difference before/after the pass-through.

## before
Screenshot 2024-05-19 at 11 17 50 PM

## after
Screenshot 2024-05-19 at 11 16 04 PM

**Pull request merged**